# This notebook has codes to using Linear Regression from scikit learn.<br>This notebook was done for an assignment for a data science class.


# Part A



In [17]:

import numpy as np

import pandas as pd

debt_pred_credit_df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Credit.csv")





# Part B



In [18]:

debt_pred_credit_df.head()


Income  Limit  Rating  Cards  Age  Education  Married  Balance
0   14.891   3606     283      2   34         11        1      333
1  106.025   6645     483      3   82         15        1      903
2  104.593   7075     514      4   71         11        0      580
3  148.924   9504     681      3   36         11        0      964
4   55.882   4897     357      2   68         16        1      331


# Part C





In [19]:



from sklearn import preprocessing

debt_pred_list_of_feature_cols = list(debt_pred_credit_df.columns)

debt_pred_list_of_feature_cols.remove('Balance')

DPCFM = debt_pred_credit_df[debt_pred_list_of_feature_cols]

Target_V = debt_pred_credit_df['Balance']

DPFCM_scaled = preprocessing.scale(DPCFM)






# Part D



In [20]:


from sklearn.model_selection import train_test_split


DPFCMS_training, DPFCMS_testing, Target_training, Target_testing = train_test_split(DPFCM_scaled, Target_V, test_size = 0.24, random_state = 9)





# Part E



In [21]:

from sklearn.linear_model import LinearRegression

dp_linear_reg = LinearRegression()

dp_linear_reg.fit(DPFCMS_training, Target_training)

dp_linear_reg_const_intercept = dp_linear_reg.intercept_

dp_liner_reg_coefficients = dp_linear_reg.coef_

print (dp_linear_reg_const_intercept) 

print (dp_liner_reg_coefficients) 


print (DPCFM.columns)




520.0391034603759
[-264.69553256  360.5601294   252.73642988   24.91707959  -17.23897686
   11.65810466   -9.82859046]
Index(['Income', 'Limit', 'Rating', 'Cards', 'Age', 'Education', 'Married'], dtype='object')



# Part E Results

Based on the results in part E, assuming the columns of the scaled feature matrix and  <br>
the original dataframe both in the same order: <br>

y = 520.0391 - 264.6955 x Income + 360.56 x Limit + 252.7364 x Rating + 24.917 x Cards - 17.2389 x Age <br>
     11.6581 x Education - 9.8285 x Married <br>
    
is our predictive model given by linear regression.





# Part F 




In [22]:

Target_prediction = dp_linear_reg.predict(DPFCMS_testing)


from sklearn import metrics
import numpy as np


mean_sq_err = metrics.mean_squared_error(Target_testing, Target_prediction)

root_mean_sq_err = np.sqrt(mean_sq_err)

print ('The root mean square error of the prediction is: ' + str(root_mean_sq_err))



The root mean square error of the prediction is: 143.2464973964227



# Part F Results
From running cell F at this moment.
The root mean square error of the prediction is: 143.2464973964227




# Part G





In [23]:



import numpy as np

import pandas as pd

debt_pred_credit_df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Credit.csv")


from sklearn import preprocessing

debt_pred_list_of_feature_cols = list(debt_pred_credit_df.columns)


debt_pred_list_of_feature_cols.remove('Balance')


DPCFM = debt_pred_credit_df[debt_pred_list_of_feature_cols]

Target_V = debt_pred_credit_df['Balance']


DPFCM_scaled = preprocessing.scale(DPCFM)


from sklearn.linear_model import LinearRegression

debt_pred_linear_reg = LinearRegression()


from sklearn.model_selection import cross_val_score

k_folds = 10

neg_mean_sq_err_list = cross_val_score(debt_pred_linear_reg,DPFCM_scaled,Target_V, cv = k_folds, scoring= 'neg_mean_squared_error')


pos_mean_sq_err_list = -neg_mean_sq_err_list

rt_mn_sq_list = np.sqrt (pos_mean_sq_err_list)

real_rmse = rt_mn_sq_list.mean()

print (real_rmse) 





160.33198910744133



# Part G Results

Based on running Part G at this moment, the RMSE is: 160.33198910744133
